### DISCUSSION TITLE
Is it possible to get System.Type from TypeDefinition?

### DISCUSSION THREAD

**Discussion by jcmrva** [jbevain/cecil#849] [Q&A] [answered]
I'm trying to get a list of types defined in an assembly. `Assembly.Load` fails when attempting to load referenced packages.

Cecil can read the assembly without difficulty but there doesn't seem to be a method to get a `System.Type` from the results.

**Comment by m-carrasco** [comment]
From my short experience, I understand that Cecil has no such feature. There are Importers (System.Type to TypeReference) but not the other way around. 

Could this be related to what you want to achieve? https://stackoverflow.com/questions/40018991/how-to-implement-isassignablefrom-with-mono-cecil

I think they suggest using these elements to build an equality relation between System.Type and TypeReferences:

> The MetadataToken seems to only be unique within the same assembly. I had the problem with two completely different types from different assemblies having the same MetadataToken. My suggestion is to also check equality for the full name. 

Best regards,
Manuel

**Reply by m-carrasco** [reply]
Another useful resource: https://groups.google.com/g/mono-cecil/c/YaDSaqHC0H8

**Reply by jcmrva** [reply]
Thank you for looking into it. The SO discussion is an interesting read; unfortunately in my case, I know which assembly the types are in (all of them).

I need to think about the other one but it probably makes more sense to refactor so this isn't necessary.

**Reply by Zastai** [reply]
I believe the `TypeReference` enables you to get the assembly-qualified name of a type, which you can use with `Type.GetType(string)` to load the defining assembly and get the type from that.

Of course, there is no guarantee that this will work, especially not if the assembly you're inspecting uses a different target framework from the running. You might then try to get a type from `mscorlib` instead of `System.Private.CoreLib` or vice versa. And just because Cecil is able to resolve the reference does not necessarily mean it's somewhere the .NET runtime will be able to find it (plus its dependencies, as applicable).

So for your specific case: if an assembly's dependencies are not where the runtime can find them, you can't get a `System.Type` from it. But if you know where they are (or can ask the user to tell you), you can just use an `AssemblyResolve` handler and use `Assembly.Load` & co.

**Comment by jbevain** [comment, marked as answer]
System.Type is implemented at the runtime level. To get a proper type the runtime needs to load a proper assembly. If you can't load an assembly you can't get a `System.Type` for types in that assembly I'm afraid.
